mediator/esmFlds.F90

```diff
@@ -94,205 +187,216 @@ subroutine med_fldList_AddFld(flds, stdname, shortname)
+    character(len=*)             , intent(in), optional :: mrg_fracname
+    integer                      , intent(out), optional :: rc
+
+    call med_FldList_addMrg(fldListTo(index)%fields, fldname, mrg_from, mrg_fld, mrg_type, mrg_fracname)
+
+  end subroutine med_fldList_AddMrgTo
+
+  !================================================================================
+
+  subroutine med_fldList_AddMrg(flds, fldname, mrg_from, mrg_fld, mrg_type, mrg_fracname)
+
+    ! ----------------------------------------------
+    ! Determine mrg entry or entries in flds aray
+    ! ----------------------------------------------
 
     ! input/output variables
-    type(med_fldList_entry_type) , pointer              :: flds(:)
+    type(med_fldList_entry_type) , intent(in), target   :: flds
     character(len=*)             , intent(in)           :: fldname
     integer                      , intent(in)           :: mrg_from
     character(len=*)             , intent(in)           :: mrg_fld
     character(len=*)             , intent(in)           :: mrg_type
     character(len=*)             , intent(in), optional :: mrg_fracname
 
     ! local variables
-    integer :: n, id
+    integer :: rc
+    type(med_fldList_entry_type), pointer :: newfld
     character(len=*), parameter :: subname='(med_fldList_AddMrg)'
     ! ----------------------------------------------
+        
+    newfld => med_fldList_GetFld(flds, fldname, rc)
+    if (chkerr(rc,__LINE__,u_FILE_u)) return    
+    newfld%merge_fields(mrg_from) = mrg_fld
+    newfld%merge_types(mrg_from) = mrg_type
+    if (present(mrg_fracname)) then
+       newfld%merge_fracnames(mrg_from) = mrg_fracname
+    end if
 
-    id = 0
-    do n= 1,size(flds)
-       if (trim(fldname) == trim(flds(n)%stdname)) then
-          id = n
-          exit
-       end if
-    end do
-    if (id == 0) then
-       do n = 1,size(flds)
-          write(6,*) trim(subname)//' input flds entry is ',trim(flds(n)%stdname)
+  end subroutine med_fldList_AddMrg
+
+  function med_fldList_GetFld(fields, fldname, rc) result(newfld)
+    use ESMF, only : ESMF_LogWrite, ESMF_END_ABORT, ESMF_LOGMSG_ERROR, ESMF_Finalize, ESMF_LOGMSG_INFO
+
+    type(med_fldList_entry_type) , intent(in), target :: fields
+    character(len=*)                  , intent(in)    :: fldname
+
+    type(med_fldList_entry_type), pointer :: newfld
+    logical :: found
+    integer :: rc
+    character(len=*), parameter :: subname='(med_fldList_GetFld)'
+    
+
+    call med_fldList_findName(fields, fldname, found, newfld)
+
+    rc = ESMF_SUCCESS
+    if(.not. found) then
+       rc = ESMF_FAILURE
+       newfld => fields
+       do while(associated(newfld))
+          write(6,*) trim(subname)//' input flds entry is ',trim(newfld%stdname)
+          newfld => newfld%next
        end do
        call ESMF_LogWrite(subname // 'ERROR: fldname '// trim(fldname) // ' not found in input flds', ESMF_LOGMSG_ERROR)
```

Comment:
Our convention is to post an error to the PET log and return to the calling function, this is done all over cesm, why do it differently here?

Reply:
Because the calling function might not have a return code that takes you to the top. So at times you need to call ESMF_Finalize directly. It would be good to check in this case.

Reply:
I put the finalize back